coral-common/src/main/java/com/linkedin/coral/common/transformers/OperatorTransformer.java

```diff
@@ -0,0 +1,75 @@
+/**
+ * Copyright 2017-2023 LinkedIn Corporation. All rights reserved.
+ * Licensed under the BSD-2 Clause license.
+ * See LICENSE in the project root for license information.
+ */
+package com.linkedin.coral.common.transformers;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlNodeList;
+import org.apache.calcite.sql.SqlSelect;
+import org.apache.calcite.sql.validate.SqlValidator;
+
+
+/**
+ * Abstract class for generic transformations on SqlNodes
+ */
+public abstract class OperatorTransformer {
+  protected SqlNode inputSqlNode;
+  protected SqlValidator sqlValidator;
+  protected List<SqlSelect> topSelectNodes = new ArrayList<>();
+
+  protected abstract boolean condition();
+
+  protected abstract SqlNode transform();
+
+  public SqlNode apply(SqlNode inputSqlNode) {
+    return apply(inputSqlNode, null);
+  }
+
+  /**
+   * @param inputSqlNode input SqlNode to be transformed
+   * @param sqlValidator Calcite SqlValidator, which is used to derive RelDataType of the SqlNode
+   * @return transformed SqlNode
+   */
+  public SqlNode apply(SqlNode inputSqlNode, SqlValidator sqlValidator) {
+    this.inputSqlNode = inputSqlNode;
+    this.sqlValidator = sqlValidator;
+    if (inputSqlNode instanceof SqlSelect) {
+      // Store the top select SqlNodes, which will be used to construct the dummy SqlSelect to derive the RelDataType
+      // of a SqlNode
+      this.topSelectNodes.add((SqlSelect) inputSqlNode);
+    }
+    if (condition()) {
+      return transform();
+    } else {
+      return inputSqlNode;
+    }
+  }
+
+  /**
+   * Get the RelDataType of the input SqlNode
+   */
+  protected RelDataType getRelDataType(SqlNode sqlNode) {
+    if (sqlValidator == null) {
+      throw new RuntimeException("Please provide sqlValidator to get the RelDataType of a SqlNode!");
+    }
+    // Traverse the stored top select SqlNodes from new to old to construct the dummy SqlSelect,
+    // return the SqlNode's RelDataType directly if SqlValidator can validate the dummy SqlSelect
+    for (int i = topSelectNodes.size() - 1; i >= 0; --i) {
+      final SqlSelect topSelectNode = topSelectNodes.get(i);
+      final SqlSelect dummySqlSelect = new SqlSelect(topSelectNode.getParserPosition(), null, SqlNodeList.of(sqlNode),
```

Comment:
(1) What are the criteria for manufacturing this dummySelectNode? Why do we selectively use "topSelectNode.getFrom()" but not the other clauses? The sqlValidator constructed like this may not be the right SOT.
instead of creating the dummySelectNode, can we do the following - 

//take the input SqlNode (SqlCall) and construct a dummyRelNode
RelNode dummyRelNode = hiveToRelConverter.toRel(sqlCall);

//extract the dummy validator (will need to make the method public)
SqlValidator validator = hiveToRelConverter.getSqlValidator();

//get the top-level scope from the top level selectNode - 
SqlValidatorScope topLevelScope = validator.getSelectScope((SqlSelect) sqlCall);

//derive the type
final RelDataType argType = validator.deriveType(topLevelScope, sqlNode);

(2) topSelectNodes is a list and we want to iterate over the entire list to identify the relDataType. However, we return the first relDataType discovered. That is always going to be derived using topSelectNodes[0]? Then why is  topSelectNodes a list?
Could you explain the process of deriving the data type here?

Reply:
(1) I used `topSelectNode.getFrom()` only because that is the only information required to construct the minimum dummy SqlSelect: `SELECT <SqlNode to get the datatype> FROM xxx`. Other parts (orderby/groupby/...) are not necessary to derive the datatype.

And as for `RelNode dummyRelNode = hiveToRelConverter.toRel(sqlCall);`, I think `SqlCall` should be the top level selectNode instead of the [inputSqlNode](https://github.com/linkedin/coral/pull/336/files#diff-2fb1eb21bc5936ee6b8406a4c9e238c018ac121a14cdc928e8e05a7677fb9887R22) ? Otherwise `toRel` will fail.

Comparing these 2 approaches of getting datatype, I think both of them work for deriving the datatype, but using SqlValidator + dummy Select would require less resources, because constructing a `ToRelConverter` requires more things apart from `SqlValidator`, and converting the whole top level selectNode would take more resources than validating the minimum dummy SqlSelect.

(2) To begin with, I only stored the last visited topSelectNode to construct the dummy SqlNode, but I found a corner case in the regression test. For example, in the following SQL:
```
SELECT *
FROM (
   SELECT * FROM db.tbl
)
LATERAL JOIN EXPLODE(col1) t AS a
WHERE t.a > 0
```
If I want to derive the datatype of `t.a`, the last visited top level SqlNode is `SELECT * FROM db.tbl`, however, `t.a` doesn't exist in `db.tbl` so it would throw exception. 
So we need to store all the top level SqlSelects (both inner `SELECT * FROM db.tbl` and the whole SQL) in the list and traverse them from end to beginning, return the datatype directly once the datatype can be derived without exception.
I will add more contexts in the comment of this method to elaborate its logic.

Reply:
(1) I used `topSelectNode.getFrom()` only because that is the only necessary information required to construct the minimum dummy SqlSelect: `SELECT <SqlNode to get the datatype> FROM <topSelectNode.getFrom()>`. Other parts (orderby/groupby/...) are not needed to derive the datatype.

I have implemented the approach that you mentioned to get the datatype, but there are some regressions on the prod views. Besides, using SqlValidator + dummy sqlSelect would require less resources, because constructing a `ToRelConverter` requires more things apart from SqlValidator, and converting the whole top level selectNode would take more resources than validating the minimum dummy SqlSelect.

(2) I have added the [detailed comment](https://github.com/linkedin/coral/pull/336/files#diff-2fb1eb21bc5936ee6b8406a4c9e238c018ac121a14cdc928e8e05a7677fb9887R50) to the method, PTAL.